spring-kafka/src/main/java/org/springframework/kafka/support/converter/ProjectingMessageConverter.java

```diff
@@ -0,0 +1,121 @@
+import org.springframework.kafka.support.KafkaNull;
+import org.springframework.messaging.Message;
+import org.springframework.util.Assert;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.jayway.jsonpath.spi.mapper.JacksonMappingProvider;
+
+/**
+ * A {@link MessageConverter} implementation that uses a Spring Data {@link ProjectionFactory} to bind incoming messages
+ * to projection interfaces.
+ *
+ * @author Oliver Gierke
+ * @since 2.1.1
+ */
+public class ProjectingMessageConverter extends MessagingMessageConverter {
+
+	private final ProjectionFactory projectionFactory;
+	private final MessagingMessageConverter delegate;
+
+	/**
+	 * Creates a new {@link ProjectingMessageConverter} using the given {@link ObjectMapper}.
+	 *
+	 * @param mapper must not be {@literal null}.
+	 */
+	public ProjectingMessageConverter(ObjectMapper mapper) {
+
+		Assert.notNull(mapper, "ObjectMapper must not be null!");
+
+		JacksonMappingProvider provider = new JacksonMappingProvider(mapper);
+		JsonProjectingMethodInterceptorFactory interceptorFactory = new JsonProjectingMethodInterceptorFactory(provider);
+
+		SpelAwareProxyProjectionFactory factory = new SpelAwareProxyProjectionFactory();
+		factory.registerMethodInvokerFactory(interceptorFactory);
+
+		this.projectionFactory = factory;
+		this.delegate = new StringJsonMessageConverter(mapper);
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see org.springframework.kafka.support.converter.MessagingMessageConverter#convertPayload(org.springframework.messaging.Message)
+	 */
+	@Override
+	protected Object convertPayload(Message<?> message) {
+		return this.delegate.convertPayload(message);
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see org.springframework.kafka.support.converter.MessagingMessageConverter#extractAndConvertValue(org.apache.kafka.clients.consumer.ConsumerRecord, java.lang.reflect.Type)
+	 */
+	@Override
+	protected Object extractAndConvertValue(ConsumerRecord<?, ?> record, Type type) {
+
+		Object value = record.value();
+
+		if (value == null) {
+			return KafkaNull.INSTANCE;
+		}
+
+		Class<?> rawType = ResolvableType.forType(type).resolve(Object.class);
+
+		if (!rawType.isInterface()) {
+			return this.delegate.extractAndConvertValue(record, type);
+		}
+
+		InputStream inputStream = new ByteArrayInputStream(getAsByteArray(value));
+
+		return this.projectionFactory.createProjection(rawType, inputStream);
+	}
+
+	/**
+	 * Returns the given source value as byte array.
+	 *
+	 * @param source must not be {@literal null}.
+	 * @return the source instance as byte array.
+	 */
+	private static byte[] getAsByteArray(Object source) {
+
+		Assert.notNull(source, "Source must not be null!");
+
+		if (source instanceof String) {
+			return String.class.cast(source).getBytes(StandardCharsets.UTF_8);
+		}
+
+		if (source instanceof byte[]) {
+			return byte[].class.cast(source);
+		}
+
+		throw new ConversionException(String.format("Unsupported payload type %s!", source.getClass()), null);
```

Comment:
Shouldn't we say in this message what types do we support? Just to be as friendly as possible for end-users of our lib.